### ISSUE TITLE
Crash when opening location picker

### ISSUE THREAD

**Issue by dmfs** [schedjoules/android-event-discovery-sdk#344]
> I get the following crash when selecting the ALL LOCATIONS item on top. So do several users.
With 1.9 I could not repoduce this. Sometime it doesn't crash though - no idea yet on why.
It shows the ENTER LOCATION/MY LOCATION for about 500ms before crashing.
Any ideas?
```
java.lang.RuntimeException: android.os.TransactionTooLargeException: data parcel size 927544 bytes
	at android.app.ActivityThread$StopInfo.run(ActivityThread.java:3776)
	at android.os.Handler.handleCallback(Handler.java:751)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:154)
	at android.app.ActivityThread.main(ActivityThread.java:6123)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:867)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:757)
Caused by: android.os.TransactionTooLargeException: data parcel size 927544 bytes
	at android.os.BinderProxy.transactNative(Native Method)
	at android.os.BinderProxy.transact(Binder.java:615)
	at android.app.ActivityManagerProxy.activityStopped(ActivityManagerNative.java:3700)
	at android.app.ActivityThread$StopInfo.run(ActivityThread.java:3768)
	... 7 more
```

**Comment by dmfs** [comment]
I think the problem is a large amount of data in the saved state of the activity.
The quick fix is probably to not store the results in the saved state and let it reload the list on configuration changes.

**Comment by dmfs** [comment]
After looking into the code it seems it's only the initial page which is retained. So another solution could be to let the EventListFragment take care of loading the initial events or to reduce the number of events on the initial result page.

**Comment by lemonboston** [comment]
I am checking this. We don't actually save the first `ResultPage` explicitly but since it is a `Fragment` argument the framework saves it to be able to restore the `Fragment`.

It's strange that it happens when opening the location picker, I assume that it means that the device is low on memory and Activities get killed instantly when they are left. The same as the "Don't keep Activities" developer setting.

These are those conditions that somehow can't be tested, simulated with current version, I am looking into that again. I could make the activity kill happen with 1.9 schedjoules-app (not demo app), so I think it's either having the `MainActivity` of the demo app that changes things or the switch from `EventListActivity` to `MicroFragmentHostActivity`. I'll keep trying to figure this out because without being able to test these conditions, it's harder to notice and fix them.

About the solution, I think we should omit the event description texts when downloading the list because I think they represent the big majority of the data. And fetch them when we load the details screen (by uid).
It's not just a quick thought, I thought about it when we discussed whether it's good if `savedInstanceState` is good for saving data relating to this:
https://developer.android.com/topic/libraries/architecture/viewmodel.html#viewmodel_vs_savedinstancestate
(@dmfs You asked then how much data is our event list.)
I think that 500 KB, which is the presumed limit, should be always enough to store everything that needs to be displayed on a screen, not counting pictures or any media of course, those are handled in different way. So only text.
The reason why we can hit that limit here is because we have the event descriptions downloaded to the list and those can be pretty long, but we don't actually show them on this screen.
So I think matching the API calls to the needs of the screens can be a solution here, like we wouldn't necessarily need the API return the event descriptions as part of the list, and the Actions and Details by UID call could be merged to have one call for the details. Just mentioning this to make clearer what I mean, but we don't necessarily need API change to make this change on our side, we could ignore descriptions and have the Actions and EventByUID call parallel when loading details for now.

@dmfs What do you think?
As this example shows, this is not just about saving data but passing them as well to Fragments, which puts the loading MFs approach in question if we wanted to refrain from putting any data except state or ids into args as Google suggests for savedInstanceState. So I think the question is this assumption of whether 500KB will always be enough for the textual content (not sure if this limit is for one component or the whole app though). 

To get a clearer picture on the amount of data, @rutgerg do you know how many events are returned for the first discovery page? (without location and date specified). Is that a fix number or can it change?

**Comment by lemonboston** [comment]
Demo app 1.9 version with "Don't keep Activities" on destroys `EventListActivity` when opening location picker, demo app 1.10 version doesn't kill its corresponding `MicroFragmentHostActivity`.
I don't know why, but I also noticed that `onSaveInstanceState()` is still called nevertheless every time. I see now in the docs
> This method is called before an activity **may** be killed..
So now I assume it's probably called for each Activity which is put to stack.

So I still don't have a way to test and reproduce these situations, but I move on now to see how much effort is make those changes I mentioned, to ignore description and load it when details is opened.

**Comment by mattlaabs** [comment]
One more detail: it seems to only happen with Android 7.x - I see 25 reports and most of them high end devices.

**Comment by lemonboston** [comment]
I've implemented a first version of the mentioned changes on branch `bugs/344-transactiontoolarge-ignore-desc`. Event descriptions are now ignored in the discovery result list and event is fetched again by uid when loading the details screen. 50 events are cached for 15 minutes to open the details quickly the second time, just like we currently do. (Can be of course changed to whatever we see fit.)

@dmfs If you happen to have time to give an opinion on this and in case it's confirmative, I could finish this tomorrow (I would move the description ignoring to java-api-client and review all changes.)

**Comment by lemonboston** [comment]
@mattlaabs Thanks for the info, that's useful, even though I tried now with 7.0 and 7.1.1 emulators and I couldn't reproduce it with those either.

**Comment by mattlaabs** [comment]
even with 68bc967087d09f91f97a44144e6de308b955b557 I can still reproduce this (now on a Nexus 5x w/ Android 8) with a data parcel size of just over 1MB. Maybe try to set the country to Germany for testing.

**Comment by mattlaabs** [comment]
same with 94c7bd3412b34d1521b5e5fa9dd1b83d06c873db ...
Here's a list of devices (mine + customers) where it is reported from:
Nexus 5x (Android O Beta)
HTC 10
Moto G(4) (athene)
OnePlus3 (OnePlus3)
Xperia X Compact (F5321)
Galaxy S8+ (dream2lte)
Honor 8 (HWFRD)
LG G6 (lucye)
Galaxy S7 Edge (hero2lte)
Galaxy S6 (zerofltevzw)

**Comment by dmfs** [comment]
I can sort of reproduce it with the 7.1.1. emulator. It doesn't crash but I see the following log messages:

```
!!! FAILED BINDER TRANSACTION !!!  (parcel size = 797284)
App sent too much data in instance state, so it was ignored
android.os.TransactionTooLargeException: data parcel size 797284 bytes
 at android.os.BinderProxy.transactNative(Native Method)
 at android.os.BinderProxy.transact(Binder.java:615)
 at android.app.ActivityManagerProxy.activityStopped(ActivityManagerNative.java:3636)
 at android.app.ActivityThread$StopInfo.run(ActivityThread.java:3773)
 at android.os.Handler.handleCallback(Handler.java:751)
 at android.os.Handler.dispatchMessage(Handler.java:95)
 at android.os.Looper.loop(Looper.java:154)
 at android.app.ActivityThread.main(ActivityThread.java:6119)
 at java.lang.reflect.Method.invoke(Native Method)
 at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
 at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
```

This is based on 1.10.1

**Comment by dmfs** [comment]
I did some debugging and watched the parcel size as it's been written. There are two big chunks written to the parcel:

1) EventListMicroFragment ~400kb
2) Fragment saved state ~ 400kb

The first one certainly can be reduced by not storing descriptions, but I'm not sure what the content of the Fragment saved state is yet. I'll look into that.

**Comment by dmfs** [comment]
The Fragment saved state contains the result page once again, because `EventListListShowFragment` gets it as a parameter. That means, the entire result page is serialized twice, once with the `EventListMicroFragment` and once with `EventListListShowFragment`.

**Comment by dmfs** [comment]
@lemonboston I like the idea of an `EventService`. I think we should take it to the next level and offload the discovery requests to this service. It should have a method which takes an `EventDiscovery` object and returns something like a result page (including prev and next page queries), which has envelopes without any event data at all (just etag and uid).
Whenever the ui would need an event, it would call the method to retrieve the event by uid.

That way, the entire event data would be stored in the service and only the etags and uids need to be parcelled. Access to individual events would still be reasonable fast, because (since the service is running in the same process) no IPC is involved.

**Comment by lemonboston** [comment]
I can see the error in the logs as well with 7.1.1 emulator, that's great.

So this is what I've seen and done:
The size was 1147772 bytes with current master in my case. With the bug fix branch ignoring the descriptions it was 627276 bytes, so it has about halved it, but still above the limit (as @mattlaabs you mentionied as well that you could reproduce it with that version).
I removed our first loader MF as suggested and loading is done now in `EventListListLoaderFragment` which passes the first page to `EventListListShowFragment` (through `EventListFragment`), so only 1 Fragment needs to store it instead of 2.
This supposedely halved the size again and I indeed stopped seeing the error. So it's around 314 KB at this point.
I've also set the result limit to 50 events which halves the default 100. So this should make it around 160 KB.
So I think we are safely below the 500 KB limit. (Since we ignore the description the variation between individual event data sizes cannot be too big, especially to have a big difference when summed over the size of 50.)
I've also added the discovery call as well to `EventService` based on the suggestion, which has a nice benefit that it can cache the fetched events individually by uid. So there is no extra network call now during loading details, it serves it from cache. (The descriptions are still ignored when returning the `ResultPage`s but they are kept in the event cache.)
I assume that caching by uid for 15 minutes is fine and we don't need to use the etags.

@dmfs I have general concerns about swiching to async loading in our show MFs in view of the actual idea as well (of keeping ids and etags and offloading to service). I will expand on it in detail on Slack.

Because of those concerns, I think we could go with this fix now, I am going to submit it as pull request. 

@mattlaabs when you reproduce the crash, is that with your app integrating with the mentioned commits of the SDK or is that the demo app of this SDK repository?
(I tried with German, no crash with that either.)

**Comment by lemonboston** [comment]
One more note: when I tried to use a single loader MF for the details for both cases of cached and not cached (actions and event), I had trouble with the transitions. The `BottomUp` one didn't finish before gathering from cache and moving on with `Revealed` so there was a big glitch. That's why I set that 300ms delay in `EventDetailLoaderMicroFragment` and also kept the routing in `BasicEventDetails` to have different route when cached data is available.
`EventDetailLoaderMicroFragment` could be used for both but that means the details page doesn't appear that immediately as in our current version.
So basically we can't really handle fast and non-fast loader MFs transparently, at least with transitions that when fired quickly after each other cause glitch.

**Comment by mattlaabs** [comment]
@lemonboston only with my app

**Comment by mattlaabs** [comment]
1.10.2 seems to fix it! Thanks!

**Comment by mattlaabs** [comment]
Unfortunately with 1.14 this problem is back. (Pixel 2, Android 8)

`java.lang.RuntimeException: android.os.TransactionTooLargeException: data parcel size 570720 bytes`

**Comment by mattlaabs** [comment]
The versions with Android arch comp also caused several problems, so I'm probably back to 1.9 until this is fixed.

**Comment by dmfs** [comment]
We'll check that.

**Comment by mattlaabs** [comment]
Also this is happening when leaving an event by pressing SAVE EVENT.

**Comment by lemonboston** [comment]
I can reproduce it, it doesn't crash for me, but shows the error in the logs.
Steps: open demo app with "10 days from now", tap on location
Device: Pixel Emulator API 26 Android 8.0

With these debug logs added to `BaseFragment`:
```java
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        Investigator.log(this, "savedState", savedInstanceState);
        Investigator.log(this, "args", getArguments());
    }


    @Override
    public void onSaveInstanceState(Bundle outState)
    {
        super.onSaveInstanceState(outState);
        Investigator.log(this, "outState", outState);
        Investigator.log(this, "args", getArguments());
    }
```

I see this:
```
11-20 17:24:35.946 D/Investigator: [main] LoaderFragment{b475d17 #0 id=0x7f0d0009}.onCreate() | savedState = null
11-20 17:24:36.010 D/Investigator: [main] LoaderFragment{b475d17 #0 id=0x7f0d0009}.onCreate() | args = Bundle[{MICRO_FRAGMENT_ENVIRONMENT=org.dmfs.android.microfragments.BasicMicroFragmentEnvironment@289d504}]
11-20 17:24:42.945 D/Investigator: [main] EventListFragment{bee7f01 #1 id=0x7f0d0009}.onCreate() | savedState = null
11-20 17:24:42.957 D/Investigator: [main] EventListFragment{bee7f01 #1 id=0x7f0d0009}.onCreate() | args = Bundle[{MICRO_FRAGMENT_ENVIRONMENT=org.dmfs.android.microfragments.BasicMicroFragmentEnvironment@b0491a6}]
11-20 17:24:43.047 D/Investigator: [main] EventListHeaderFragment{f5643f5 #0 id=0x7f0d00dc}.onCreate() | savedState = null
11-20 17:24:43.057 D/Investigator: [main] EventListHeaderFragment{f5643f5 #0 id=0x7f0d00dc}.onCreate() | args = Bundle[{schedjoules.RELOAD_EVENT_LIST_CAGE=com.schedjoules.eventdiscovery.framework.serialization.boxes.ParcelableBox@4edc58a}]
11-20 17:24:43.084 D/Investigator: [main] FilterFragment{3ab78d7 #1 id=0x7f0d00dd}.onCreate() | savedState = null
11-20 17:24:43.094 D/Investigator: [main] FilterFragment{3ab78d7 #1 id=0x7f0d00dd}.onCreate() | args = Bundle[{schedjoules.CATEGORIES_CAGE=com.schedjoules.eventdiscovery.framework.serialization.boxes.ParcelableBox@c1e25c4}]
11-20 17:24:43.373 D/Investigator: [main] EventListListShowFragment{e811465 #2 id=0x7f0d00de}.onCreate() | savedState = null
11-20 17:24:43.384 D/Investigator: [main] EventListListShowFragment{e811465 #2 id=0x7f0d00de}.onCreate() | args = Bundle[{schedjoules.EVENTS_RESULT_PAGE=com.schedjoules.eventdiscovery.framework.serialization.boxes.EventResultPageBox@c4aa53a}]
11-20 17:25:17.819 D/Investigator: [main] LocationPickerFragment{a8d2f5b #0 id=0x1020002}.onCreate() | savedState = null
11-20 17:25:17.829 D/Investigator: [main] LocationPickerFragment{a8d2f5b #0 id=0x1020002}.onCreate() | args = null
11-20 17:25:19.403 D/Investigator: [main] EventListFragment{bee7f01 #1 id=0x7f0d0009}.onSaveInstanceState() | outState = Bundle[{schedjoules.FILTER_CATEGORIES=com.schedjoules.eventdiscovery.framework.serialization.boxes.IterableBox@58f0a34}]
11-20 17:25:19.422 D/Investigator: [main] EventListFragment{bee7f01 #1 id=0x7f0d0009}.onSaveInstanceState() | args = Bundle[{MICRO_FRAGMENT_ENVIRONMENT=org.dmfs.android.microfragments.BasicMicroFragmentEnvironment@b0491a6}]
11-20 17:25:19.440 D/Investigator: [main] EventListHeaderFragment{f5643f5 #0 id=0x7f0d00dc}.onSaveInstanceState() | outState = Bundle[{}]
11-20 17:25:19.457 D/Investigator: [main] EventListHeaderFragment{f5643f5 #0 id=0x7f0d00dc}.onSaveInstanceState() | args = Bundle[{schedjoules.RELOAD_EVENT_LIST_CAGE=com.schedjoules.eventdiscovery.framework.serialization.boxes.ParcelableBox@4edc58a}]
11-20 17:25:19.471 D/Investigator: [main] FilterFragment{3ab78d7 #1 id=0x7f0d00dd}.onSaveInstanceState() | outState = Bundle[{}]
11-20 17:25:19.487 D/Investigator: [main] FilterFragment{3ab78d7 #1 id=0x7f0d00dd}.onSaveInstanceState() | args = Bundle[{schedjoules.CATEGORIES_CAGE=com.schedjoules.eventdiscovery.framework.serialization.boxes.ParcelableBox@c1e25c4}]
11-20 17:25:19.506 D/Investigator: [main] EventListListShowFragment{e811465 #2 id=0x7f0d00de}.onSaveInstanceState() | outState = Bundle[{}]
11-20 17:25:19.523 D/Investigator: [main] EventListListShowFragment{e811465 #2 id=0x7f0d00de}.onSaveInstanceState() | args = Bundle[{schedjoules.EVENTS_RESULT_PAGE=com.schedjoules.eventdiscovery.framework.serialization.boxes.EventResultPageBox@c4aa53a}]

11-20 17:25:21.767 E/JavaBinder: !!! FAILED BINDER TRANSACTION !!!  (parcel size = 625192)
11-20 17:25:21.774 E/ActivityThread: App sent too much data in instance state, so it was ignored
                                     android.os.TransactionTooLargeException: data parcel size 625192 bytes
                                         at android.os.BinderProxy.transactNative(Native Method)
                                         at android.os.BinderProxy.transact(Binder.java:748)
                                         at android.app.IActivityManager$Stub$Proxy.activityStopped(IActivityManager.java:4636)
                                         at android.app.ActivityThread$StopInfo.run(ActivityThread.java:3998)
                                         at android.os.Handler.handleCallback(Handler.java:789)
                                         at android.os.Handler.dispatchMessage(Handler.java:98)
                                         at android.os.Looper.loop(Looper.java:164)
                                         at android.app.ActivityThread.main(ActivityThread.java:6541)
                                         at java.lang.reflect.Method.invoke(Native Method)
                                         at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
```

So the event list is not actually saved in `savedInstanceState` explicitly anywhere (only categories in `EventListFragment`, but that's not much data), so I assume now that it's the `Fragment`s argument bundle that gets saved as well by the framework (I can't find it in framework code), and that causes the exception.
`EventListListShowFragment` has the `ResultPage` events in the arguments, so I am going to look into that now how can we avoid that, for example whether we could replace those 3 fragments (loading, show, error) with custom views, or make `EventListFragment` manage all that itself.

**Comment by lemonboston** [comment]
I've found the problem why this size has gone up again above the limit: I've removed the result page limit request parameter accidentally during a refactor. It had been set to 50 instead of the default 100 to halve the size of the data.
I am going to set that to 40 now to have a hotfix for this.

The reported error had size `570720`, the ones I encountered ranged from `572532` to `625192` so getting 40% of that will make sure to stay below the `500000` limit.

**Comment by lemonboston** [comment]
I could also reproduce the `TransactionTooLargeException` when leaving the details screen with the SHARE action (size was `538164`). That's the same thing, so reducing the result page limit fixes that, too (tested).

Note, that besides this fix, we should of course get back to this topic of replacing this way of passing on the event list data with a better solution.

**Comment by dmfs** [comment]
@mattlaabs 1.14.1 should work around this issue. It's not the final solution yet. Though, it should buy us enough time to approach the underlying issue with a solid fix.
